### PR TITLE
Add support for charging weapons

### PR DESCRIPTION
**Pull request by exinfinitum**
This PR adds the following:

- `chargeTime` attribute for weapons, which denotes time in frames that a weapon must charge before firing (power and ship counter are consumed, but ammo and weapon capacitor/temperature are not consumed during charging)
- `chargeEffect` and `chargeSound` to denote effect and sound used when charging
- `playSoundOncePerBurst` - if `True`, weapon only plays firing sound on the first shot in a burst
- `onChargeWeapon` effect that functions similarly to `onFireWeapon` but only is fired while the weapon is charging
- `chargeTime`, `fireCharge`, `fireRepeat` and `repeating` fields for `fireEffects` and `chargeEffects` - note that `fireCharge` will be equal to `chargeTime` when weapon is done charging and is actually charging, and `fireRepeat` will be zero when weapon is charging

Note that `onFireWeapon` and other weapon fire events are not triggered when a weapon is charging.

### CODE REVIEW

File: Mammoth/TSE/CWeaponClass.cpp [hunk 1]
```diff
@@ -2420,6 +2506,58 @@ bool CWeaponClass::FireWeapon (CInstalledDevice &Device,
 	if (Shots.GetCount() == 0)
 		return false;
 
+	//	If we're charging, then we don't fire shots - instead we only increment the ship counter (if needed),
+	//	create the fire effect, and return True (so that we consume power). Charging does not consume ammo,
+	//	and cannot fail due to damaged or disrupted weapons.
+	if (IsCharging)
+		{
```
Comment (gmoromisato) [inline Mammoth/TSE/CWeaponClass.cpp:2513]: Probably should move this to a function and call it from here. Something like:

if (IsCharging)
   {
   return ChargeWeapon(...)
   }

That will keep FireWeapon cleaner

**Review by gmoromisato — CHANGES_REQUESTED** [review]
Looks pretty good. I made a few local comments; shouldn't take too long to change.

File: Mammoth/Include/TSEDevices.h [hunk 2]
```diff
@@ -244,7 +244,7 @@ class CDeviceClass
 				}
 
 			CTargetList &GetTargetList ();
-			void SetTargetList (CTargetList &TargetListArg)
+			void SetTargetList (CTargetList TargetListArg)
```
[the remark this answers is not in the document]
Comment (exinfinitum) [reply Mammoth/Include/TSEDevices.h:247]: Seems like this was a mistake that happened when I was updating the PR with the latest master branch. Fixed in latest revision.

**Comment by exinfinitum** [comment]
This MR should now be ready for review again.

**Review by gmoromisato — APPROVED** [review]

**Pull request merged**